### DISCUSSION TITLE
Development | Updates - 21/02/21

### DISCUSSION THREAD

**Discussion by johanjanssens** [joomlatools/joomlatools-pages#551] [Announcements]
A little update on the work done in the past week. 

### 1. Custom field improvements

Worked with @kbrookes on improving the custom field support to make it easier to grab fields for a specific field group and to get a flatten array of all the field values.

You can retrieve the fields for a specific group by using the new `group([name])` method

`<? $group = collection()->fields->group('[name]') ?>`

You can retrieve an associative array of field values using the new `values()` method

`<? $values = collection()->fields->group('[name]')->values() ?>`

For more info see: https://github.com/joomlatools/joomlatools-pages/pull/363

### 2. Layout improvements

Based on a question from @kbrookes  worked on making it easier to get and set the layout properties in a scenario where you want to set layout properties dynamically for example by retrieving them from a custom field group. You can find the complete discussion here: https://github.com/joomlatools/joomlatools-pages/discussions/545

To make this easier, together with the custom field improvements reworked the layout property handling as part of https://github.com/joomlatools/joomlatools-pages/pull/540 Layout properties are now recursively loaded and merged before the page is rendered, this gives you  access to all layout properties in the page and in any layout, or partial. 

For example, consider you have, a page using a article layout that extends from the default layout, each defining different layout properties:

#### /pages/page.html.php

```php
---
layout:
    path: article
    options:
        new: defined_in_page
---

<?= layout()->options); ?>
```

#### /layouts/article.hmtl.php

```php
---
layout: default
options:
     bar: defined_in_article_layout
---

<ktml:content>
```

#### /layouts/default.hmtl.php

```php
---
options:
   foo: defined_in_default_layout
   bar: defined_in_default_layout
---

<ktml:content>
```

The result would be

```json

{
   "new": "defined_in_page",
   "bar": "defined_in_article_layout", 
   "foo": "defined_in_default_layout"
}
```

Also added a new `layout()`method to make it easier to get the layout properties, it returns a config object and when cast to a string will output JSON. 

More changes for the layout property handling are planned as part of 0.20 release see also following issue https://github.com/joomlatools/joomlatools-pages/issues/547

### 3. Released v0.19.6

The v0.19.6 release is out with various improvements and bug fixes. More info: https://github.com/joomlatools/joomlatools-pages/releases/tag/v0.19.6

Happy coding!

**Comment by kbrookes** [comment]
Lots of toys for me in this one :)
